src/System.Private.CoreLib/src/System/Threading/LowLevelThread.cs

```diff
@@ -64,30 +62,77 @@ internal static unsafe int WaitForMultipleObjects(IntPtr* pHandles, int numHandl
-                        throw ex;
+                    /// Check for duplicate handles. This is a brute force O(n^2) search, which is intended since the typical
+                    /// array length is short enough that this would actually be faster than using a hash set. Also, the worst
+                    /// case is not so bad considering that the array length is limited by
+                    /// <see cref="WaitHandle.MaxWaitHandles"/>.
+                    for (int i = 1; i < numHandles; ++i)
+                    {
+                        IntPtr handle = pHandles[i];
+                        for (int j = 0; j < i; ++j)
+                        {
+                            if (pHandles[j] == handle)
+                            {
+                                throw new DuplicateWaitObjectException("waitHandles[" + i + ']');
+                            }
+                        }
+                    }
                 }
+
+                ThrowWaitFailedException(errorCode);
             }
 
             return result;
         }
 
+        internal static void ThrowWaitFailedException(int errorCode)
+        {
+            switch (errorCode)
+            {
+                case Interop.mincore.Errors.ERROR_INVALID_HANDLE:
+                    throw InvalidOperationException.NewInvalidHandle();
+
+                case Interop.mincore.Errors.ERROR_INVALID_PARAMETER:
+                    throw new ArgumentException();
+
+                case Interop.mincore.Errors.ERROR_ACCESS_DENIED:
+                    throw new UnauthorizedAccessException();
+
+                case Interop.mincore.Errors.ERROR_NOT_ENOUGH_MEMORY:
+                    throw new OutOfMemoryException();
+
+                case Interop.mincore.Errors.ERROR_TOO_MANY_POSTS:
+                    /// Only applicable to <see cref="WaitHandle.SignalAndWait(WaitHandle, WaitHandle)"/>. Note however, that
+                    /// if the semahpore already has the maximum signal count, the Windows SignalObjectAndWait function does not
+                    /// return an error, but this code is kept for historical reasons and to convey the intent, since ideally,
+                    /// that should be an error.
+                    throw new InvalidOperationException(SR.Threading_SemaphoreFullException);
+
+                case Interop.mincore.Errors.ERROR_NOT_OWNER:
+                    /// Only applicable to <see cref="WaitHandle.SignalAndWait(WaitHandle, WaitHandle)"/> when signaling a mutex
+                    /// that is locked by a different thread. Note that if the mutex is already unlocked, the Windows
+                    /// SignalObjectAndWait function does not return an error.
+                    throw new SynchronizationLockException();
+                    // TODO: netstandard2.0 - After switching to ns2.0 contracts, use the below instead for compatibility
+                    //throw new ApplicationException(SR.Arg_SynchronizationLockException);
```

Comment:
Looks like it's there in S.P.CoreLib but I can't test for it yet because it's not yet in the API surface used by the tests